platform-dependent-components/problem-solver/cxx/verification-module/manager/duplications_check_manager.cpp

```diff
@@ -0,0 +1,382 @@
+
+  for (auto const & connector : connectors)
+  {
+    bool insertedOnPreviousStep = insertedOnCurrentStep;
+    ScIterator3Ptr relationIterator = m_context->CreateIterator3(relationType, ScType::ConstPermPosArc, connector);
+
+    bool arcContainsRelations = false;
+    while (relationIterator->Next())
+    {
+      arcContainsRelations = true;
+      auto const & insertionResult = foundRelations.insert(relationIterator->Get(0));
+      if (!insertionResult.second)
+        relationsInfo.duplicatingRelations.insert(relationIterator->Get(0));
+      else
+      {
+        relationsInfo.containsConnectorsWithoutIntersectingRelations = insertedOnPreviousStep;
+        insertedOnCurrentStep = true;
+      }
+    }
+
+    relationsInfo.containsConnectorsWithoutRelations |= !arcContainsRelations;
+  }
+}
+
+void DuplicationsCheckManager::CheckSingularRelations(
+    ScAddr const & checkedElement,
+    ElementDuplicationsCheckResult & checkResult) const
+{
+  // presumably there are less singular relations than connectors, outgoing from the checked element
+  for (auto const & relation : singularRelations)
+  {
+    ScType const & relationType = m_context->GetElementType(relation);
+    ScType arcType = relationType == ScType::ConstNodeRole ? ScType::ConstPermPosArc : ScType::ConstCommonArc;
+
+    ScIterator5Ptr relationPairsIterator =
+        m_context->CreateIterator5(checkedElement, arcType, ScType::Unknown, ScType::ConstPermPosArc, relation);
+
+    unsigned relationPairsNum = 0;
+    while (relationPairsIterator->Next())
+    {
+      relationPairsNum++;
+      if (relationPairsNum > 1)
+      {
+        checkResult.errorsDescriptions.emplace_back(
+            "Duplicate relation " + IdentifierUtils::GetIdentifiersString(m_context, relation) + " from "
+            + IdentifierUtils::GetIdentifiersString(m_context, checkedElement)
+            + ". Expected only one outgoing relation pair per element.");
+
+        break;
+      }
+    }
+  }
+}
+
+void DuplicationsCheckManager::CheckQuasybinaryRelations(
+    ScAddr const & checkedElement,
+    ElementDuplicationsCheckResult & checkResult) const
+{
+  auto processRelationPair = [this](
+                                 ScAddr const & relation,
+                                 ScAddr const & tuple,
+                                 ScAddrToValueUnorderedMap<ScAddrList> & foundQuasybinaryRelationsTuples) -> void
+  {
+    if (!quasybinaryRelations.count(relation))
+      return;
+
+    foundQuasybinaryRelationsTuples[relation].push_back(tuple);
+  };
+
+  auto checkRelationsSets = [this](
+                                ScAddrToValueUnorderedMap<ScAddrList> const & foundQuasybinaryRelationsTuples,
+                                ElementDuplicationsCheckResult & checkResult) -> void
+  {
+    for (auto const & relationWithTuple : foundQuasybinaryRelationsTuples)
+      CheckDuplicationInQuasybinaryRelationSets(relationWithTuple.first, relationWithTuple.second, checkResult);
+  };
```

Comment:
since you've decided to use lambdas they can take checkResult and foundQuasubinary... map from context as references
if you declare map before lambdas
I'm not sure that it'll improve something, just that 2 variables aren't changing inside cycles